ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -139,3 +159,85 @@ muxChannel pmss mid md w cnt =
            then pure Nothing
            else return $ Just blob
 
+muxBearerNew :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
+             => (MuxSDU ptcl -> m (Time m)) -- Write action
+             -> m (MuxSDU ptcl, Time m)     -- Read action
+             -> m Word16                    -- SDU size action
+             -> m ()                        -- Close action
+             -> m (MuxBearer ptcl m)
+muxBearerNew wfn rwn sfn cfn = do
+    st <- atomically $ newTVar Larval
+    return $ MuxBearer wfn rwn sfn cfn st
+
+muxBearerSetState :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
+                  => MuxBearer ptcl m
+                  -> MuxBearerState
+                  -> m ()
+muxBearerSetState bearer newState = atomically $ writeTVar (state bearer) newState
+
+
+muxClient :: (MonadAsync m, MonadFork m, MonadSay m, MonadSTM m, MonadThrow m,
+            Ord ptcl, Enum ptcl, Bounded ptcl, HasCallStack)
+        => [(Version, MiniProtocolDescriptions ptcl m)]
+        -> MuxBearer ptcl m
+        -> m (Version, MiniProtocolDescriptions ptcl m)
+muxClient verMpds bearer = do
+    let versions = map fst verMpds
+        msg = MsgInitReq versions
+        blob = toLazyByteString $ encodeCtrlMsg msg
+        sdu = MuxSDU (RemoteClockModel 0) Muxcontrol ModeInitiator (fromIntegral $ BL.length blob) blob
+
+    void $ write bearer sdu
+    (rsp, _) <- Ouroboros.Network.Mux.Types.read bearer
+    if msId rsp /= Muxcontrol || msMode rsp /= ModeResponder
+       then throwM $ MuxError MuxUnknownMiniProtocol "invalid muxInit rsp id or mode" callStack
+       else do
+           let rspMsg_e = CBOR.deserialiseFromBytes decodeCtrlMsg (msBlob rsp)
+           case rspMsg_e of
+                Left e -> throwM e
+                Right (_, MsgInitRsp version) ->
+                    -- verify that rsp version matches one of our proposals
+                    case lookup version verMpds of
+                         Nothing   -> error "muxInit invalid version selected"
```

Comment:
Should we throw a proper exception here?

Reply:
Definitely, exceptions and testcases are missing in a few places still.